libpod/container_api.go

```diff
@@ -125,84 +166,80 @@ func (c *Container) Start(ctx context.Context) (err error) {
+	if err = c.checkDependenciesAndHandleError(ctx); err != nil {
+		return err
 	}
 
+	return c.restartWithTimeout(ctx, timeout)
+}
+
+// helper function that handles dependencies and prints a helpful message
+func (c *Container) checkDependenciesAndHandleError(ctx context.Context) error {
 	notRunning, err := c.checkDependenciesRunning()
 	if err != nil {
-		return nil, errors.Wrapf(err, "error checking dependencies for container %s", c.ID())
+		return errors.Wrapf(err, "error checking dependencies for container %s", c.ID())
 	}
 	if len(notRunning) > 0 {
 		depString := strings.Join(notRunning, ",")
-		return nil, errors.Wrapf(ErrCtrStateInvalid, "some dependencies of container %s are not started: %s", c.ID(), depString)
+		return errors.Wrapf(ErrCtrStateInvalid, "some dependencies of container %s are not started: %s", c.ID(), depString)
 	}
 
-	defer func() {
-		if err != nil {
-			if err2 := c.cleanup(ctx); err2 != nil {
-				logrus.Errorf("error cleaning up container %s: %v", c.ID(), err2)
-			}
-		}
-	}()
+	return nil
+}
 
-	if err := c.prepare(); err != nil {
-		return nil, err
+func (c *Container) startDependencies(ctx context.Context) error {
+	depCtrIDs := c.Dependencies()
+	if len(depCtrIDs) == 0 {
+		return nil
 	}
 
-	if c.state.State == ContainerStateStopped {
-		// Reinitialize the container if we need to
-		if err := c.reinit(ctx); err != nil {
-			return nil, err
-		}
-	} else if c.state.State == ContainerStateConfigured ||
-		c.state.State == ContainerStateExited {
-		// Or initialize it if necessary
-		if err := c.init(ctx); err != nil {
-			return nil, err
+	depCtrs := make([]*Container, 0) //, len(depCtrIDs))
+	for _, ctrID := range depCtrIDs {
+		ctr, err := c.runtime.state.LookupContainer(ctrID)
+		if err != nil {
+			return errors.Wrapf(err, "error starting dependency for container %s. container %s not found", c.ID(), ctrID)
 		}
+		depCtrs = append(depCtrs, ctr)
 	}
 
-	attachChan := make(chan error)
+	// Build a dependency graph of containers
+	graph, err := buildContainerGraph(depCtrs)
+	if err != nil {
+		return errors.Wrapf(err, "error generating dependency graph for container %s", c.ID())
+	}
 
-	// Attach to the container before starting it
-	go func() {
-		if err := c.attach(streams, keys, resize, true); err != nil {
-			attachChan <- err
-		}
-		close(attachChan)
-	}()
+	ctrErrors := make(map[string]error)
+	ctrsVisited := make(map[string]bool)
 
-	return attachChan, nil
+	// If there are no containers without dependencies, we can't start
+	// Error out
+	if len(graph.noDepNodes) == 0 {
+		return errors.Wrapf(ErrNoSuchCtr, "All dependencies have dependencies of %s", c.ID())
+	}
+
+	// Traverse the graph beginning at nodes with no dependencies
+	for _, node := range graph.noDepNodes {
+		startNode(ctx, node, false, ctrErrors, ctrsVisited, true)
+	}
+
+	if len(ctrErrors) > 0 {
+		return errors.Wrapf(ErrCtrExists, "error stopping some containers")
```

Comment:
We should make some sort of attempt to get the errors out here - maybe logrus.Errorf them and then return?

Also, swap ErrCtrExists to... probably ErrInternal is better?